src/peergos/server/space/JdbcUsageStore.java

```diff
@@ -0,0 +1,370 @@
+                throw new RuntimeException(sqe);
+            }
+            return new UserUsage(totalBytes, errored, pending);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private int getUserId(String username) {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT id FROM users WHERE name = ?;")) {
+            insert.setString(1, username);
+            ResultSet resultSet = insert.executeQuery();
+            resultSet.next();
+            return resultSet.getInt(1);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private int getWriterId(PublicKeyHash writer) {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT id FROM writers WHERE key_hash = ?;")) {
+            insert.setBytes(1, writer.toBytes());
+            ResultSet resultSet = insert.executeQuery();
+            resultSet.next();
+            return resultSet.getInt(1);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public void addWriter(String owner, PublicKeyHash writer) {
+        try (PreparedStatement insert = conn.prepareStatement(commands.insertOrIgnoreCommand("INSERT ", "INTO writers (key_hash) VALUES(?)"))) {
+            insert.setBytes(1, writer.toBytes());
+            insert.executeUpdate();
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+        int userId = getUserId(owner);
+        int writerId = getWriterId(writer);
+
+        try (PreparedStatement insert = conn.prepareStatement(commands.insertOrIgnoreCommand("INSERT ", "INTO writerusage (writer_id, user_id, direct_size) VALUES(?, ?, ?)"))) {
+            insert.setInt(1, writerId);
+            insert.setInt(2, userId);
+            insert.setInt(3, 0);
+            insert.executeUpdate();
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public Set<PublicKeyHash> getAllWriters() {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT key_hash FROM writers;")) {
+            Set<PublicKeyHash> res = new HashSet<>();
+            ResultSet resultSet = insert.executeQuery();
+            while (resultSet.next())
+                res.add(PublicKeyHash.decode(resultSet.getBytes(1)));
+            return res;
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private String getOwner(PublicKeyHash writer) {
```

Comment:
sql statement can join writerusage and users tables

Reply:
yes please :-)

Reply:
like this:  SELECT u.name FROM users u, writerusage wu WHERE u.id = wu.user_id AND wu.writer_id = 1

Reply:
thank you!